### PR TITLE
group email validation code along with unit test

### PR DESCRIPTION
**Pull request by ssranjani06**
Fixes https://github.com/vinyldns/vinyldns/issues/1208 

Changes in this pull request:
-We have added emailvalidation  for groups in the api side

### CODE REVIEW

**Review by Aravindh-Raju — CHANGES_REQUESTED** [review]
Add the config details in [config-api.md](https://github.com/vinyldns/vinyldns/blob/master/modules/docs/src/main/mdoc/operator/config-api.md) file too.

File: modules/api/src/test/scala/vinyldns/api/domain/membership/MembershipServiceSpec.scala [hunk 1]
```diff
@@ -282,6 +285,11 @@ class MembershipServiceSpec
         verify(mockMembershipRepo, never())
           .saveMembers(any[DB], anyString, any[Set[String]], isAdmin = anyBoolean)
       }
+
+      "return an error if an invalid email is entered" in {
+        val error = underTest.createGroup(groupInfo.copy(email = "test@ok.com"), okAuth).value.unsafeRunSync().swap.toOption.get
+        error shouldBe a[EmailValidationError]
+      }
```
Comment (Aravindh-Raju) [inline modules/api/src/test/scala/vinyldns/api/domain/membership/MembershipServiceSpec.scala:292]: Can you add a similar unit test to verify wildcard email domain (eg: `*dummy.com`)?

You can change the following line in this file to include the wildcard email domain:
```
 private val mockValidEmailConfig = ValidEmailConfig(valid_domains = List("test.com", "*dummy.com"))
 ```

**Review by Aravindh-Raju — APPROVED** [review]

**Review by Jay07GIT — CHANGES_REQUESTED** [review]

File: modules/api/src/test/functional/tests/membership/create_group_test.py [hunk 2]
```diff
@@ -10,7 +10,7 @@ def test_create_group_success(shared_zone_test_context):
 
     try:
         new_group = {
-            "name": f"test-create-group-success{shared_zone_test_context.partition_id}",
```
Comment (Jay07GIT) [inline modules/api/src/test/functional/tests/membership/create_group_test.py:13]: I am curious to know, why this string interpolation was removed?.

File: modules/api/src/main/scala/vinyldns/api/domain/membership/MembershipService.scala [hunk 3]
```diff
@@ -364,6 +369,27 @@ class MembershipService(
...
   }.toResult
 
+  def emailValidation(email: String): Result[Unit] = {
+    val emailDomains = validDomains.valid_domains
+    val emailRegex = if(emailDomains.isEmpty){
+      """^[a-zA-Z0-9\.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$""".r
+    }
+    else {
+      if (emailDomains.mkString(",").contains("*")) {
+        ("^[A-Za-z0-9._%+-]+@" + emailDomains.mkString(",").replaceAllLiterally("*", "[A-Za-z0-9.]*").replaceAllLiterally(",", "|") + "$").r
```
Comment (Jay07GIT) [inline modules/api/src/main/scala/vinyldns/api/domain/membership/MembershipService.scala:379]: Hope this condition is checking the wildcard in entire domain string. If yes, this should validate the string starts with wildcard, not with the entire string.

File: modules/api/src/main/scala/vinyldns/api/domain/membership/MembershipService.scala [hunk 4]
```diff
@@ -364,6 +369,27 @@ class MembershipService(
     }
   }.toResult
 
+  def emailValidation(email: String): Result[Unit] = {
+    val emailDomains = validDomains.valid_domains
+    val emailRegex = if(emailDomains.isEmpty){
+      """^[a-zA-Z0-9\.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$""".r
+    }
+    else {
+      if (emailDomains.mkString(",").contains("*")) {
```
Comment (Jay07GIT) [inline modules/api/src/main/scala/vinyldns/api/domain/membership/MembershipService.scala:378]: wild card can be * and %. Kindly add condition for both.

**Review by Jay07GIT — CHANGES_REQUESTED** [review]

**Review by arpit4ever — CHANGES_REQUESTED** [review]

**Review by nspadaccino — CHANGES_REQUESTED** [review]
I found some edge cases we are not currently handling:

```
test@.test.com
test@..test.com
test@.@.test.com
test@.@@.test.com
test@....@@@.test.com
```

For the above, it looks like characters between the local and domain parts aren't being checked properly.

`test@test.com.test.com`

It looks like you can chain domain parts together with dots.

```
.test@test.com
te.....st@test.com
```

It is currently allowing dots at the beginning of the local part, and also consecutive dots.

```
te@st@test.com
@te@st@test.com
```

We are currently allowing @ symbols in the local part. There should only be one @ symbol allowed.

`tes"(),:;<>@[\]t@test.com`

These special characters should not be allowed. According to RFC 5322, the only special characters that should be allowed are 

> !#$%&'*+-/=?^_`{|}~

This Wikipedia article details the RFC email address requirements with a bunch of examples: 
https://en.wikipedia.org/wiki/Email_address#Syntax

I also found a useful article detailing a variety of different regexes for different levels of email validation: 
https://www.baeldung.com/java-email-validation-regex

**Comment by nspadaccino** [comment]
New regex looks good, but can you add some unit tests for some of the examples I gave in my previous comment?

File: modules/api/src/test/scala/vinyldns/api/domain/membership/MembershipServiceSpec.scala [hunk 5]
```diff
@@ -282,8 +298,145 @@ class MembershipServiceSpec
         verify(mockMembershipRepo, never())
           .saveMembers(any[DB], anyString, any[Set[String]], isAdmin = anyBoolean)
       }
-    }
 
+      "return an error if an invalid domain is entered" in {
+        val error = underTest.createGroup(groupInfo.copy(email = "test@ok.com"), okAuth).value.unsafeRunSync().swap.toOption.get
+        error shouldBe a[emailValidationError]
+      }
```
Comment (Aravindh-Raju) [inline modules/api/src/test/scala/vinyldns/api/domain/membership/MembershipServiceSpec.scala:305]: We can have new line after each test here too following the existing file format

File: modules/api/src/test/scala/vinyldns/api/domain/membership/MembershipServiceSpec.scala [hunk 6]
```diff
@@ -29,12 +29,15 @@ import vinyldns.core.domain.auth.AuthPrincipal
...
 import scalikejdbc.{ConnectionPool, DB}
+import vinyldns.api.config.ValidEmailConfig
 import vinyldns.api.domain.zone.NotAuthorizedError
 import vinyldns.core.TestMembershipData._
 import vinyldns.core.TestZoneData._
 import vinyldns.core.domain.membership._
 import vinyldns.core.domain.record.RecordSetRepository
 
+
+
```
Comment (Aravindh-Raju) [inline modules/api/src/test/scala/vinyldns/api/domain/membership/MembershipServiceSpec.scala:40]: Can remove this added new lines.

File: modules/core/src/main/scala/vinyldns/core/Messages.scala [hunk 7]
```diff
@@ -81,4 +81,8 @@ object Messages {
 
   // Error displayed when group name or email is empty
   val GroupValidationErrorMsg = "Group name and email cannot be empty."
+
+  val emailValidationErrorMsg = "Please enter a valid Email ID. Valid domains should end with"
+
+  val invalidEmailValidationErrorMsg = "Please enter a valid Email ID."
```
Comment (Aravindh-Raju) [inline modules/core/src/main/scala/vinyldns/core/Messages.scala:87]: As mentioned already, we can capitalise the first letters here.

**Review by Aravindh-Raju — CHANGES_REQUESTED** [review]

**Comment by ssranjani06** [comment]
> New regex looks good, but can you add some unit tests for some of the examples I gave in my previous comment?



> New regex looks good, but can you add some unit tests for some of the examples I gave in my previous comment?



> New regex looks good, but can you add some unit tests for some of the examples I gave in my previous comment?



> 
I added unit tests as mentioned. I also made a change in the regex where the email will not start with dot and end with dot.
It will not allow two consecutive dots.

File: modules/api/src/test/functional/tests/membership/create_group_test.py [hunk 8]
```diff
@@ -114,7 +114,21 @@ def test_create_group_without_name_or_email(shared_zone_test_context):
...
 
+        new_group = {
+            "name": "invalid-email",
+            "email": "test@abc.com",
+            "description": "this is a description",
+            "members": [{"id": "ok"}],
+            "admins": [{"id": "ok"}]
+        }
+        error = client.create_group(new_group, status=400)
+        assert_that(error, is_("Please enter a valid Email ID.Valid domains should end with test.com,dummy.com"))
```
[the remark this answers is not in the document]
Comment (ssranjani06) [reply modules/api/src/test/functional/tests/membership/create_group_test.py:167]: added

**Review by arpit4ever — APPROVED** [review]
LGTM
Verified locally.

**Review by nspadaccino — APPROVED** [review]
